File: UsingFMDB-Objective-C/UsingFMDB-Objective-CTests/BookCacheTests.m
//
//  BookCacheTests.m
//  UsingFMDB-Objective-C
//
//  Created by akabeko on 2016/12/26.
//  Copyright © 2016年 akabeko. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "Book.h"
#import "BookCache.h"

/**
 * Test the book data.
 */
@interface BookCacheTests : XCTestCase

@end

@implementation BookCacheTests

/**
 * Setup the test instance.
 */
- (void)setUp {
    [super setUp];
}

/**
 * Tear down the test instance.
 */
- (void)tearDown {
    [super tearDown];
}

/**
 * Test the additional data processing.
 */
- (void)testAdd {
    BookCache *cache = [[BookCache alloc] init];
    Book      *book  = [Book bookWithId:kBookIdNone author:@"" title:@"" releaseDate:[NSDate date]];
    XCTAssert(!([cache add:book]));
    XCTAssertEqual([cache.authors count], 0);

    book = [Book bookWithId:1 author:@"author" title:@"title" releaseDate:[NSDate date]];
    XCTAssert([cache add:book]);
    XCTAssertEqual(cache.authors.count, 1);

    NSArray *books = [cache.booksByAuthor objectForKey:@"author"];
    XCTAssertEqual(books.count, 1);

    book = [books objectAtIndex:0];
    XCTAssertEqualObjects(book.title, @"title");
}

/**
 * Test the remove data processing.
 */
- (void)testRemove {
    BookCache *cache = [[BookCache alloc] init];
    Book      *book  = [Book bookWithId:1 author:@"author" title:@"title" releaseDate:[NSDate date]];
    XCTAssert([cache add:book]);
    XCTAssertEqual(cache.authors.count, 1);
    XCTAssert([cache remove:book]);
    XCTAssertEqual(cache.authors.count, 0);

    NSArray *books = [cache.booksByAuthor objectForKey:@"author"];
    XCTAssertNil(books);
}

/**
 * Test the update data with title processing.
 */
- (void)testUpdateTitle {
    BookCache *cache = [[BookCache alloc] init];
    Book      *book1 = [Book bookWithId:1 author:@"author" title:@"title" releaseDate:[NSDate date]];
    XCTAssert([cache add:book1]);
    XCTAssertEqual([cache.authors count], 1);

    Book     *book2 = [Book bookWithId:1 author:@"author" title:@"title2" releaseDate:[NSDate date]];
    XCTAssert([cache update:book1 newBook:book2]);
    XCTAssertEqual(cache.authors.count, 1);

    NSArray *books = [cache.booksByAuthor objectForKey:@"author"];
    XCTAssertEqual(books.count, 1);

    book2 = [books objectAtIndex:0];
    XCTAssertEqualObjects(book2.title, @"title2");
}

/**
 * Test the update data with author and title processing.
 */
- (void)testUpdateAuthorAndTitle {
    BookCache *cache = [[BookCache alloc] init];
    Book      *book1 = [Book bookWithId:1 author:@"author" title:@"title" releaseDate:[NSDate date]];
    XCTAssert([cache add:book1]);
    XCTAssertEqual(cache.authors.count, 1);

    Book     *book2 = [Book bookWithId:1 author:@"author2" title:@"title2" releaseDate:[NSDate date]];
    XCTAssert([cache update:book1 newBook:book2]);
    XCTAssertEqual(cache.authors.count, 1);

    NSArray *books = [cache.booksByAuthor objectForKey:@"author"];
    XCTAssertNil(books);

    books = [cache.booksByAuthor objectForKey:@"author2"];
    XCTAssertEqual(books.count, 1);

    book2 = [books objectAtIndex:0];
    XCTAssertEqualObjects(book2.author, @"author2");
    XCTAssertEqualObjects(book2.title, @"title2");
}

@end

File: UsingFMDB-Swift/UsingFMDB-SwiftTests/BookCacheTests.swift
//
//  BookCacheTests.swift
//  UsingFMDB-Swift
//
//  Created by akabeko on 2016/12/22.
//  Copyright © 2016年 akabeko. All rights reserved.
//

import XCTest

/// Test the book data.
class BookCacheTests: XCTestCase {
    /// Setup the test instance.
    override func setUp() {
        super.setUp()
    }

    /// Tear down the test instance.
    override func tearDown() {
        super.tearDown()
    }

    /// Test the additional data processing.
    func testAdd() {
        let cache = BookCache()
        var book  = Book(bookId: Book.BookIdNone, author: "", title: "", releaseDate: Date())
        XCTAssert(!cache.add(book: book))
        XCTAssertEqual(cache.authors.count, 0)

        book = Book(bookId: 1, author: "author", title: "title", releaseDate: Date())
        XCTAssert(cache.add(book: book))
        XCTAssertEqual(cache.authors.count, 1)

        let books = cache.booksByAuthor["author"]
        XCTAssertEqual(books?.count, 1)
        XCTAssertEqual(books?[0].title, "title")
    }

    /// Test the remove data processing.
    func testRemove() {
        let cache = BookCache()
        let book  = Book(bookId: 1, author: "author", title: "title", releaseDate: Date())
        XCTAssert(cache.add(book: book))
        XCTAssertEqual(cache.authors.count, 1)
        XCTAssert(cache.remove(book: book))
        XCTAssertEqual(cache.authors.count, 0)

        let books = cache.booksByAuthor["author"]
        XCTAssertNil(books)

        let book2  = Book(bookId: 2, author: "author", title: "title", releaseDate: Date())
        XCTAssert(cache.add(book: book))
        XCTAssert(cache.add(book: book2))
        XCTAssert(cache.remove(book: book))
        XCTAssertEqual(cache.authors.count, 1)
    }

    /// Test the update data with title processing.
    func testUpdateTitle() {
        let cache = BookCache()
        let book1 = Book(bookId: 1, author: "author", title: "title", releaseDate: Date())
        XCTAssert(cache.add(book: book1))
        XCTAssertEqual(cache.authors.count, 1)

        let book2 = Book(bookId: 1, author: "author", title: "title2", releaseDate: Date())
        XCTAssert(cache.update(oldBook: book1, newBook: book2))
        XCTAssertEqual(cache.authors.count, 1)

        var books = cache.booksByAuthor["author"]
        XCTAssertEqual(books?[0].title, "title2")
    }

    /// Test the update data with author and title processing.
    func testUpdateAuthorAndTitle() {
        let cache = BookCache()
        let book1 = Book(bookId: 1, author: "author", title: "title", releaseDate: Date())
        XCTAssert(cache.add(book: book1))
        XCTAssertEqual(cache.authors.count, 1)

        let book2 = Book(bookId: 1, author: "author2", title: "title2", releaseDate: Date())
        XCTAssert(cache.update(oldBook: book1, newBook: book2))
        XCTAssertEqual(cache.authors.count, 1)

        var books = cache.booksByAuthor["author"]
        XCTAssertNil(books)

        books = cache.booksByAuthor["author2"]
        XCTAssertEqual(books?[0].author, "author2")
        XCTAssertEqual(books?[0].title, "title2")
    }
}
